lib/using_restricted_array.rb

```diff
@@ -6,56 +6,119 @@
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) because we will need to loop through the whole array each time
+# Space complexity: O(1) because no additional memory is required because no copy is made
 def length(array)
-  raise NotImplementedError
+  my_integer_array_length = 0
+  while array[my_integer_array_length] != nil
+    my_integer_array_length += 1
+  end
+  return my_integer_array_length
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) because we will need to loop through the whole array each time
+# Space complexity: O(1) because no additional memory is required because no copy is made
 def print_array(array)
-  raise NotImplementedError
+  index = 0
+  while array[index] != nil
+    p array[index]
+    index += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) because, worst case scenerio, we need to loop through the whole array each time
+# Space complexity: O(1) because no additional memory is required because no copy is made
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  index = 0
+  until index == length
+    if array[index] == value_to_find
+      return true
+      exit
```

Comment:
⚠️ `exit` generally leaves the program, you almost certainly don't want that here

```suggestion
```